### PR TITLE
maintainers-list.nix: Document (current) invitation process, add get-maintainer.sh

### PR DESCRIPTION
**Pull request by nicoonoclaste**
## Description of changes

- Add `maintainers/scripts/get-maintainer.sh` to query `maintainers-list.nix`
  It supports querying by exact match against a maintainer's handle or any of the documented attributes of `maintainers.${handle}`.

- Document the [RFC 39] invitation process to [nixpkgs-maintainers] as it stands now.

- ~~Add new `reinvite-maintainers.yml` workflow, which:~~
  - triggers on comments on #234293, from non-bot users, that contain “invite me” ;
  - checks whether the GitHub user is present in `maintainers-list.nix` (taken from branch `master`) ;
  - invites the user to the [nixpkgs-maintainers] team, created by [RFC 39] ;
  - replies (if successful) to the user, informing them they were invited to the team.

[nixpkgs-maintainers]: https://github.com/orgs/NixOS/teams/nixpkgs-maintainers
[RFC 39]: https://github.com/NixOS/rfcs/blob/master/rfcs/0039-unprivileged-maintainer-teams.md#summary

## Things done

- Tested `get-maintainer.nix`
  - [x] all selectors work, on both existing maintainers and negative queries ;
  - [x] the script always exits with a non-zero exit code when no match was found.
  The latter allows the workflow to more-easily check whether a match exists.

- ~~`reinvite-maintainers.yml` isn't tested yet~~
  I guess the only way to do so would be to test it on a dummy org?

### CODE REVIEW

**Review by zimbatm — COMMENTED** [review]
Great initiative and the workflow makes sense. The logic will probably need to be moved to the bot that invites maintainers due to the security restrictions of GHA.

**Comment by infinisil** [comment]
I feel like there's a good chance for this to become out of sync or slightly different from the RFC 39 implementation that does the initial invites.

A self-serve page does sound a lot easier though. If that exists we could have some very simple automation that posts a comment in PRs adding to the `maintainer-list.nix`, telling people to visit the page to get invited. This way we wouldn't need the continuously-running service anymore.

A small worry is that this wouldn't conform with RFC 39 anymore, which [specified how the automation should be done](https://github.com/NixOS/rfcs/blob/master/rfcs/0039-unprivileged-maintainer-teams.md#team-automation). So to do this right it should be another RFC to override the old one. Probably not a bad idea tbh, makes sure we don't miss anything. It would be entirely uncontroversial though, so I think this could be very quick.

If you're up for the task, I think this would be great to have, a PoC along with an RFC.

**Comment by zimbatm** [comment]
@nbraud deploying an app seems better indeed. This will allow us to direct people to maintainers.nixos.org (or bikeshed later) to check their status and add themselves. The app would also either have to re-implement, or re-use the code of RFC39 so it doesn't get out of sync. If you're up for it then let's go.

File: maintainers/scripts/get-maintainer.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  exit 1
+}
+
+listAsJSON() {
+  nix-instantiate --eval --strict --json "${MAINTAINERS_DIR}/maintainer-list.nix"
+}
+
+parseArgs() {
+  [ $# -gt 0 -a $# -lt 3 ] || \
+    die "usage: '$0' [selector] value"
```
Comment (mkg20001) [inline maintainers/scripts/get-maintainer.sh:23]: maybe an example would be good

**Review by mkg20001 — APPROVED** [review]

**Review by infinisil — CHANGES_REQUESTED** [review]

File: maintainers/scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+and supporting data or nixlang files as needed.
+
+What follows is a (very incomplete) overview of available scripts.
+
+
+## Metadata
+
+### `get-maintainer.sh`
+
+`get-maintainer.sh [selector] [value]` returns a JSON object describing
```
Comment (infinisil) [inline maintainers/scripts/README.md:13]:
```suggestion
`get-maintainer.sh [selector] value` returns a JSON object describing
```

`value` is not optional

File: maintainers/scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+a given nixpkgs maintainer, equivalent to `lib.maintainers.${x} // { handle = x; }`.
+
+The maintainer is designated by a `selector` which must be one of:
+- `handle` (default): the maintainer's attribute name in `lib.maintainers`;
+- `email`, `name`, `github`, `githubId`, `matrix`, `name`:
+  attributes of the maintainer's object, matched exactly;
+  see [`maintainer-list.nix`] for the fields' definition.
+
```
Comment (infinisil) [inline maintainers/scripts/README.md:21]: An example would be really helpful here, once with the default and once with another selector

File: maintainers/scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+This folder contains various executable scripts for nixpkgs maintainers,
+and supporting data or nixlang files as needed.
+
+What follows is a (very incomplete) overview of available scripts.
+
+
+## Metadata
+
+### `get-maintainer.sh`
```
Comment (infinisil) [inline maintainers/scripts/README.md:11]: Something this is missing is the motivation for why we need this. Can you describe the use case for this script here?

**Review by infinisil — CHANGES_REQUESTED** [review]

File: maintainers/scripts/README.md (hunk 4 above)
Comment (infinisil) [reply maintainers/scripts/README.md:11]: Still good to write this down, but I just had a need for this script myself! Had to find out what the GitHub handle of somebody on Matrix was :D

File: maintainers/scripts/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Maintainer scripts
+
+This folder contains various executable scripts for nixpkgs maintainers,
+and supporting data or nixlang files as needed.
```
Comment (infinisil) [inline maintainers/scripts/README.md:4]:
```suggestion
This folder contains various executable scripts for nixpkgs maintainers,
and supporting data or nixlang files as needed.
These scripts generally aren't a stable interface and may changed or be removed.
```

I think generally files in the source shouldn't be relied upon, so just emphasising this here.

File: maintainers/scripts/README.md (hunk 2 above)
Comment (infinisil) [reply maintainers/scripts/README.md:14]: Oh I'm just suggesting to not have the `[]` brackets. That usually indicates optional fields in the CLI, whereas if you don't have them it indicates a required field.

File: maintainers/scripts/README.md (hunk 4 above)
Comment (infinisil) [reply maintainers/scripts/README.md:12]: Ah yeah you're right, but yes that text would be great to add!

**Review by infinisil — APPROVED** [review]
Awesome, this is looking great! I'm really glad we got some docs going for the maintainer scripts :heart:

**Pull request merged**